### ISSUE TITLE
Possible to reference tile using relative path?

### ISSUE THREAD

**Issue by jihu** [repaint-io/maven-tiles#31]
Hi,

I'm currently evaluating tiles for our big restructuring of our maven project, and so far it seems really good! But there is one feature that I haven't seen in the documentation, and wonder if it is possible or not.

Our tiles will all be part of the same project structure, basically with all the tiles in a tiles-subfolder in the directory of the project. And we would prefer if we didn't have to install the tiles before using them. Is this possible? Ie making the tile plugin use the files on disk (ie not a repository, not even the local on one the machine) directly? For example, using something similar to the <relativePath> tag that can be used in the <parent> section of a pom. We plan to have all our tiles as snapshots.

So, basically, we want to be able to do this:
1. Edit the tile.xml in the tile "tile1"
2. Run "mvn clean install" on another project, that references "tile1"
3. See results that reflects the changes performed in step 1

Regards
/Jimi

**Comment by jihu** [comment]
Small update: I tried creating an aggregator pom, with the test tile as first module, and the test-project as the second module. Then I tried running "mvn clean install" on the aggregator pom, but it fails with an error indicating that it can't find the artifact for the test tile in any repository. So tiles-maven-plugin tries to use the tile before it is built. Surely there must be a way to build a tile, and use it, in the same go?

/Jimi

**Comment by kdeng** [comment]
You can try add "-D" in your command line. 

For example: 

mvn -DtileGroupId:tileArtifactId=/absolute-path/tile clean install

**Comment by rvowles** [comment]
I'm not sure that even makes sense? A tile has to be defined and in your
repository (even if it is local) before use? If you could create an example
project and upload it it would make a much more useful starting point.

On Tue, Sep 15, 2015 at 6:20 AM, jihu <notifications@github.com> wrote:

> Small update: I tried creating an aggregator pom, with the test tile as
> first module, and the test-project as the second module. Then I tried
> running "mvn clean install" on the aggregator pom, but it fails with an
> error indicating that it can't find the artifact for the test tile in any
> repository. So tiles-maven-plugin tries to use the tile before it is built.
> Surely there must be a way to build a tile, and use it, in the same go?
>
> /Jimi
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/repaint-io/maven-tiles/issues/31#issuecomment-140165429>
> .
>



-- 
---
Richard Vowles,
Groovy, Java, Javascript, AngularJS

ph: +64275467747, web: www.google.com/+RichardVowles

**Comment by jihu** [comment]
Could you please explain how adding "-DtileGroupId:tileArtifactId=/absolute-path/tile" would help? Is there a built in function in maven it self? If so, can you post a link to some page describing this feature? Or is it a feature of maven-tiles?

Either way, I didn't get it to work. And the output was identical to the output without this parameter, making me suspect that the parameter wasn't picked up at all.

And even if it would work, it feels like a hack. I would prefer a solution that is capsulated inside the pom, instead of needing to add special parameters to the mvn command.

The way I see it, maven-tiles breaks one of the the base functionallity of maven. Because when a plain multi module project is built, all modules are built in the order according to "reactor sorting" (https://maven.apache.org/guides/mini/guide-multiple-modules.html). But with maven-tiles, it tries to reach an artifact without letting maven build it first, even though it should according to the aggregator pom and according to the reactor sorting algorithm.

Test case for this:

https://www.dropbox.com/s/fwy2v2n9jq9xctr/test-cases.zip?dl=0

The folder test-case1 contains a self contained test case using maven-tiles-plugin, while test-case2 contains a test case with plain maven stuff.

Test case 1:
* Go into folder test-case1
* Run mvn clean install

Expected result:
* Both modules are built and installed without failure
* An output of "Hello world" is displayed

Actual result:
* Build fails with error: "Could not find artifact mytest1:test-tile:xml:trunk-SNAPSHOT"

Workaround needed:
* Manually install test-tile first, **then** run the main pom


Test case 2:
* Go into folder test-case2
* Run mvn clean install

Expected result:
* Both modules are built and installed without failure
* An output of "Hello world" is displayed

Actuall result:
* Same as expected

**Comment by jihu** [comment]
"I'm not sure that even makes sense? A tile has to be defined and in your repository (even if it is local) before use?"

Yes, but and maven **would** have built the tile, and installed it the local repository, **if** tiles-maven-plugin wouldn't have tried to use it before maven gets a chance to do that.


"If you could create an example project and upload it it would make a much more useful starting point."

See my previous comment.

**Comment by talios** [comment]
On 15 Sep 2015, at 6:20, jihu wrote:

> Small update: I tried creating an aggregator pom, with the test tile as first module, and the test-project as the second module. Then I tried running "mvn clean install" on the aggregator pom, but it fails with an error indicating that it can't find the artifact for the test tile in any repository. So tiles-maven-plugin tries to use the tile before it is built. Surely there must be a way to build a tile, and use it, in the same go?

Unfortunately no.  Since the Maven Lifecycle is only run ONCE over the entire reactor.

For this reason, the tiles need to be installed first, otherwise they can't be resolved.

You could maybe have a pair of <profile>'s  - one with the tiles, one with the app - but that's also messy.


-- 
Mark Derricutt
http://www.theoryinpractice.net
http://www.chaliceofblood.net
http://plus.google.com/+MarkDerricutt
http://twitter.com/talios
http://facebook.com/mderricutt

**Comment by talios** [comment]
On 15 Sep 2015, at 22:53, jihu wrote:

> The way I see it, maven-tiles breaks one of the the base functionallity of maven. Because when a plain multi module project is built, all modules are built in the order according to "reactor sorting" 

You're 100% correct in that maven-tiles is effectively subverting the normal course of maven, by rewriting it's parents.

> But with maven-tiles, it tries to reach an artifact without letting maven build it first, even though it should according to the aggregator pom and according to the reactor sorting algorithm.

This is an unfortunate side effect of how the lifecycle's are managed.

It -could- be possible that we detect that the tile is in the reactor, and fork a maven lifecycle to build it, and then continue - could be something you might look at it you needed that behaviour.

Ideologically, maven-tiles is meant to free you from parents/multi-module builds, but reactor builds can still be needed/desired.

Mark


-- 
Mark Derricutt
http://www.theoryinpractice.net
http://www.chaliceofblood.net
http://plus.google.com/+MarkDerricutt
http://twitter.com/talios
http://facebook.com/mderricutt

**Comment by rvowles** [comment]
But your tiles should NOT be in them (reactor builds).

On Wed, Sep 16, 2015 at 5:36 PM, Mark Derricutt <notifications@github.com>
wrote:

> On 15 Sep 2015, at 22:53, jihu wrote:
>
> > The way I see it, maven-tiles breaks one of the the base functionallity
> of maven. Because when a plain multi module project is built, all modules
> are built in the order according to "reactor sorting"
>
> You're 100% correct in that maven-tiles is effectively subverting the
> normal course of maven, by rewriting it's parents.
>
> > But with maven-tiles, it tries to reach an artifact without letting
> maven build it first, even though it should according to the aggregator pom
> and according to the reactor sorting algorithm.
>
> This is an unfortunate side effect of how the lifecycle's are managed.
>
> It -could- be possible that we detect that the tile is in the reactor, and
> fork a maven lifecycle to build it, and then continue - could be something
> you might look at it you needed that behaviour.
>
> Ideologically, maven-tiles is meant to free you from parents/multi-module
> builds, but reactor builds can still be needed/desired.
>
> Mark
>
>
> --
> Mark Derricutt
> http://www.theoryinpractice.net
> http://www.chaliceofblood.net
> http://plus.google.com/+MarkDerricutt
> http://twitter.com/talios
> http://facebook.com/mderricutt
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/repaint-io/maven-tiles/issues/31#issuecomment-140631196>
> .
>



-- 
---
Richard Vowles,
Groovy, Java, Javascript, AngularJS

ph: +64275467747, web: www.google.com/+RichardVowles

**Comment by jihu** [comment]
OK. I think I understand the reason for this behaivor now. Thanks guys.

And I certainly see the usage of this plugin, but not for our specific use case (the tile were only to be used within the project itself). And we managed to resolve the inheritence problem that originally caused me to look into tiles. But I will definately keep tiles in mind if a more suitable use case comes up for is in the future. I really like the fundamental idea of it. :)

**Comment by talios** [comment]
On 16 Sep 2015, at 17:44, Richard Vowles wrote:

> But your tiles should NOT be in them (reactor builds).

This is true.  -SHOULD- is not -MUST- tho.

However, given that it can't actually work (reliably, from a clean, 0 release state ) it's more a "SHOULD NOT, NO, DON'T ARGUE, JUST STOP IT, DON'T. NO!" :)

-- 
Mark Derricutt
http://www.theoryinpractice.net
http://www.chaliceofblood.net
http://plus.google.com/+MarkDerricutt
http://twitter.com/talios
http://facebook.com/mderricutt
